src/lib/core/trigger.cc

```diff
@@ -54,30 +53,135 @@ trigger_fiber_f(va_list ap)
+	rlist_add_tail_entry(list, run_link, in_run);
+	rlist_add_tail_entry(&trigger->run_links, run_link, in_trigger);
+	return 0;
+}
+
+/** Take the next trigger from the run list and free the corresponding link. */
+static struct trigger *
+run_list_take_trigger(struct rlist *list)
+{
+	struct run_link *link = rlist_shift_entry(list, struct run_link,
+						  in_run);
+	struct trigger *trigger = link->trigger;
+	rlist_del_entry(link, in_trigger);
+	mempool_free(&run_link_pool, link);
+	return trigger;
+}
+
+/** Empty the run list and free all the links allocated for it. */
+static void
+run_list_clear(struct rlist *list)
+{
+	while (!rlist_empty(list))
+		run_list_take_trigger(list);
+}
+
+/** Execute the triggers in an order specified by \a list. */
+static int
+trigger_run_list(struct rlist *list, void *event)
+{
+	while (!rlist_empty(list)) {
+		struct trigger *trigger = run_list_take_trigger(list);
+		if (trigger->run(trigger, event) != 0) {
+			run_list_clear(list);
+			return -1;
+		}
+	}
+	return 0;
+}
+
 int
 trigger_run(struct rlist *list, void *event)
 {
-	struct trigger *trigger, *tmp;
-	rlist_foreach_entry_safe(trigger, list, link, tmp)
-		if (trigger->run(trigger, event) != 0)
+	/*
+	 * A list holding all triggers scheduled for execution. It's important
+	 * to save all triggers in a separate run_list and iterate over it
+	 * instead of the passed list, because the latter might change anyhow
+	 * while the triggers are run:
+	 *  * the current element might be removed from the list
+	 *    (rlist_foreach_entry_safe helps with that)
+	 *  * the next element might be removed from the list
+	 *    (rlist_foreach_entry_safe fails to handle that, but plain
+	 *    rlist_foreach_entry works just fine)
+	 *  * trigger lists might be swapped (for example see
+	 *    space_swap_triggers() in alter.cc), in which case neither
+	 *    rlist_foreach_entry nor rlist_foreach_entry_safe can help
+	 */
+	RLIST_HEAD(run_list);
+	struct trigger *trigger;
+	rlist_foreach_entry(trigger, list, link) {
+		if (run_list_put_trigger(&run_list, trigger) != 0) {
+			run_list_clear(&run_list);
 			return -1;
-	return 0;
+		}
+	}
+	return trigger_run_list(&run_list, event);
 }
 
 int
 trigger_run_reverse(struct rlist *list, void *event)
 {
-	struct trigger *trigger, *tmp;
-	rlist_foreach_entry_safe_reverse(trigger, list, link, tmp)
-		if (trigger->run(trigger, event) != 0)
+	RLIST_HEAD(run_list);
+	struct trigger *trigger;
+	rlist_foreach_entry_reverse(trigger, list, link) {
+		if (run_list_put_trigger(&run_list, trigger) != 0) {
+			run_list_clear(&run_list);
 			return -1;
-	return 0;
+		}
+	}
+	return trigger_run_list(&run_list, event);
+}
+
+void
+trigger_clear(struct trigger *trigger)
+{
+	rlist_del_entry(trigger, link);
+	struct run_link *link, *tmp;
+	rlist_foreach_entry_safe(link, &trigger->run_links, in_trigger, tmp) {
+		rlist_del_entry(link, in_run);
+		mempool_free(&run_link_pool, link);
+	}
+	rlist_create(&trigger->run_links);
```

Comment:
You don't seem to need this `rlist_create()` - an emptied list is supposed to be the same as a newly created one.

Reply:
The list is not emptied: link has two "hanldes" for two different lists: I'm only removing it from `run_list`, but not from `trigger->run_links`. This was an optimisation, I guess. To clear the run_links list at once instead of doing it element by element. Can't remember why I wrote it this way.

This might be simpler:
```diff
diff --git a/src/lib/core/trigger.cc b/src/lib/core/trigger.cc
index 6442dc7a6..8bb86a74c 100644
--- a/src/lib/core/trigger.cc
+++ b/src/lib/core/trigger.cc
@@ -173,9 +173,9 @@ trigger_clear(struct trigger *trigger)
        struct run_link *link, *tmp;
        rlist_foreach_entry_safe(link, &trigger->run_links, in_trigger, tmp) {
                rlist_del_entry(link, in_run);
+               rlist_del_entry(link, in_trigger);
                mempool_free(&run_link_pool, link);
        }
-       rlist_create(&trigger->run_links);
 }
 
 int
```